zipkin.go

```diff
@@ -1,139 +1,351 @@
+	sampled     bool
+	serviceName string
+}
+
+func newSpan(name string, traceID int64, spanId int64, parentID *int64, serviceName string) *Span {
+	zipkinSpan := &zipkincore.Span{
+		Name:              name,
+		ID:                spanId,
+		TraceID:           traceID,
+		ParentID:          parentID,
+		Annotations:       make([]*zipkincore.Annotation, 0),
+		BinaryAnnotations: make([]*zipkincore.BinaryAnnotation, 0),
 	}
 
-	tracer := &Tracer{trace.NewSpanManager(), &KafkaCollector{kafkaProducer, config.Topic}}
-	tracer.manager.Configure(config.SampleRate, false, &zipkin.Endpoint{
-		Ipv4:        config.Ip,
-		Port:        config.Port,
-		ServiceName: config.ServiceName})
-	tracer.manager.RegisterTraceCollector(tracer.collector)
-	return tracer, nil
+	return &Span{span: zipkinSpan, serviceName: serviceName}
+}
+
+func (s *Span) Sampled() bool {
+	return s.sampled
+}
+
+func (s *Span) TraceID() int64 {
+	return s.span.TraceID
+}
+
+func (s *Span) ParentID() *int64 {
+	return s.span.ParentID
 }
 
-func (t *Tracer) InitServer(spanName string, r trace.Request) context.Context {
-	ctx := t.createContextFromRequest(spanName, r)
+func (s *Span) ID() int64 {
+	return s.span.ID
+}
+
+func (s *Span) ServerReceive() {
+	log.Debugf("[Zipkin] ServerReceive")
+	s.Annotate(zipkincore.SERVER_RECV)
+}
+
+func (s *Span) ServerSend() {
+	log.Debugf("[Zipkin] ServerSend")
+	s.Annotate(zipkincore.SERVER_SEND)
+}
 
-	Annotate(ctx, zipkin.SERVER_RECV)
+func (s *Span) ServerSendAndCollect() {
+	s.ServerSend()
+	s.Collect()
+}
+
+func (s *Span) ServerReceiveAndCollect() {
+	s.ServerReceive()
+	s.Collect()
+}
+
+func (s *Span) ClientSend() {
+	log.Debugf("[Zipkin] ClientSend")
+	s.Annotate(zipkincore.CLIENT_SEND)
+}
 
-	return ctx
+func (s *Span) ClientReceive() {
+	log.Debugf("[Zipkin] ClientReceive")
+	s.Annotate(zipkincore.CLIENT_RECV)
 }
 
-func (t *Tracer) TraceClient(ctx *context.Context, spanName string) func(*error) {
-	return t.manager.TraceWithSpanNamed(ctx, spanName)
+func (s *Span) ClientReceiveAndCollect() {
+	s.ClientReceive()
+	s.Collect()
 }
 
-func (t *Tracer) TraceClientSent(ctx context.Context, spanName string) {
-	t.manager.TraceWithSpanNamed(&ctx, spanName)
-	t.CollectFromContext(ctx)
+func (s *Span) ClientSendAndCollect() {
+	s.ClientSend()
+	s.Collect()
 }
 
-func Annotate(ctx context.Context, v string) {
-	if span, ok := trace.SpanFromContext(ctx); ok {
-		span.AnnotateTimestamp(v, monotime.Now(), nil, nil)
+func (s *Span) NewChild(name string) *Span {
+	log.Debugf("[Zipkin] Creating new child span: %s", name)
+	if !s.sampled {
+		return &Span{}
 	}
+	child := newSpan(name, s.span.TraceID, newID(), &s.span.ID, s.serviceName)
+	child.collector = s.collector
+	child.ip = s.ip
+	child.port = s.port
+	child.sampled = true
+	return child
 }
 
-func (t *Tracer) Collect(s *trace.Span) {
-	if !s.TraceDisabled() {
-		data := s.Export()
-		t.collector.Collect(data)
+func (t *Tracer) NewSpanFromAvro(name string, traceInfo interface{}) *Span {
```

Comment:
interesting.. I'd consider putting something in the README about avro propagation, or a link somewhere else about how this works or what it buys you.